Add `eagerness` option description and example for `prefetch()`

Marocco2 opened the pull request:
<!-- Thank you for opening a PR! We really appreciate you taking the time to help out 🙌 -->

#### Description (required)

As per request, I've added a new section for Prefetch docs page describing what the new option does and a use case that it would be useful

<!-- Please describe the change you are proposing, and why -->

<!-- Please make changes in **one language** only -->

<!-- For a new/changed feature in an upcoming Astro release? -->
<!-- 1. Uncomment the line below, update the minor version number if known, and include a PR link -->
#### For Astro version: `5.6`. See astro PR [#13405](https://github.com/withastro/astro/pull/13405)

<!-- 2. Check that your PR includes `<p><Since v="4.x.0" /></p>` and imports the `<Since>` component, if necessary! -->

<!-- #### First-time contributor to Astro Docs? -->

<!-- If you are a member of the Astro Discord, please add your username in the description so we can welcome you there! -->
<!-- https://astro.build/chat -->

astrobot-houston commented:
<!-- lunaria-action-comment -->

## Lunaria Status Overview


🌕 **This pull request will trigger status changes.**

<details>
<summary>Learn more</summary>

By default, every PR changing files present in the [Lunaria configuration's `files` property](https://lunaria.dev/reference/configuration/#files-required) will be considered and trigger status changes accordingly.

You can change this by adding one of the keywords present in the [`ignoreKeywords` property in your Lunaria configuration file](https://lunaria.dev/reference/configuration/#ignorekeywords) in the PR's title (ignoring all files) or by [including a tracker directive](https://lunaria.dev/guides/tracking/#tracker-directives) in the merged commit's description. 
</details>

### Tracked Files




| File                                                                                                                                                        | Note                                                       |
| ----------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------- |
| [en/guides/prefetch.mdx](https://github.com/withastro/docs/blob/a777b92e85390b8b2529c342bf3b63a7d5b95a80/src%2Fcontent%2Fdocs%2Fen%2Fguides%2Fprefetch.mdx) | Source changed, localizations will be marked as outdated.  |

<details>
<summary>Warnings reference</summary>

| Icon | Description                                                                                                                                       |
| ---- | ------------------------------------------------------------------------------------------------------------------------------------------------- |
| 🔄️  | The source for this localization has been updated since the creation of this pull request, make sure all changes in the source have been applied. |

</details>

astrobot-houston commented:
Hello! Thank you for opening your **first PR** to Astro’s Docs! 🎉

Here’s what will happen next:

1. Our GitHub bots will run to check your changes.
   If they spot any broken links you will see some error messages on this PR.
   Don’t hesitate to ask any questions if you’re not sure what these mean!

2. In a few minutes, you’ll be able to see a preview of your changes on Netlify 🥳.

3. One or more of our maintainers will take a look and may ask you to make changes.
   We try to be responsive, but don’t worry if this takes a few days.

sarah11918 commented on src/content/docs/en/guides/prefetch.mdx:
```diff
@@ -114,6 +115,32 @@ To ignore slow connection detection, you can use the `ignoreSlowConnection` opti
+
+<p><Since v="5.6.0" /></p>
+
+With [`clientPrerender`](https://docs.astro.build/en/reference/experimental-flags/client-prerender/) experiment enabled, you can use `eagerness` option, following the same API described in [Speculation Rules API](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/speculationrules#eagerness).
+
+For example, you have a set of links that you want to be prerendered/prefetched but there are too many for browsers to handle (Chrome have some [limits in place](https://developer.chrome.com/blog/speculation-rules-improvements#chrome-limits)).
+To prevent that, you can set `eagerness: 'moderate'` to get advantage of FIFO strategies and browser heuristics to let it decide when prerender/prefetch them and in what order.
```
```suggestion
With the experimental [`clientPrerender`](https://docs.astro.build/en/reference/experimental-flags/client-prerender/) flag enabled, you can use the `eagerness` option on `prefetch()` to suggest to the browser how eagerly it should prefetch/prerender link targets.

This follows the same API described in [Speculation Rules API](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/speculationrules#eagerness) and defaults to `immediate` (the most eager option). In decreasing order of eagerness, the other options are `eager`, `moderate`, and `conservative`.

The `eagerness` option allows you to balance the benefit of reduced wait times against bandwidth, memory, and CPU costs for your site visitors. Some browsers, such as Chrome, have [limits in place](https://developer.chrome.com/blog/speculation-rules-improvements#chrome-limits) to guard against over-speculating (prerendering/prefetching too many links).

To use `prefetch()` programmatically with large sets of links, you can set `eagerness: 'moderate'` to get advantage of [First In, First Out (FIFO)](https://en.wikipedia.org/wiki/FIFO_(computing_and_electronics)) strategies and browser heuristics to let the browser decide when to prerender/prefetch them and in what order:
```

sarah11918 commented on src/content/docs/en/guides/prefetch.mdx:
````diff
@@ -114,6 +115,32 @@ To ignore slow connection detection, you can use the `ignoreSlowConnection` opti
 prefetch('/about', { ignoreSlowConnection: true });
 ```
 
+### Prefetch strategies with `eagerness` option
````
```suggestion
### `eagerness`
```
I think we can just use the option name here for simplicity!

sarah11918 commented on src/content/docs/en/guides/prefetch.mdx:
````diff
@@ -114,6 +115,32 @@ To ignore slow connection detection, you can use the `ignoreSlowConnection` opti
+For example, you have a set of links that you want to be prerendered/prefetched but there are too many for browsers to handle (Chrome have some [limits in place](https://developer.chrome.com/blog/speculation-rules-improvements#chrome-limits)).
+To prevent that, you can set `eagerness: 'moderate'` to get advantage of FIFO strategies and browser heuristics to let it decide when prerender/prefetch them and in what order.
+
+```astro
````
``````suggestion
```astro "{eagerness: 'moderate'}"
``````
I think this will help highlight the option!

sarah11918 reviewed (commented):
Thank you for submitting the docs for this great new feature, @Marocco2 ! I'm sure people will really appreciate having this!

I just made some editing suggestions to make this sound more like the rest of our docs and follow some of our styles. You are the expert in this topic, so please correct anything that's wrong, and make any suggestions back if something needs fixing!

Marocco2 commented:
> Thank you for submitting the docs for this great new feature, @Marocco2 ! I'm sure people will really appreciate having this!
> 
> I just made some editing suggestions to make this sound more like the rest of our docs and follow some of our styles. You are the expert in this topic, so please correct anything that's wrong, and make any suggestions back if something needs fixing!

Thank you for taking your time making the changes! I found all of them reasonable, so I've merged them 👍

sarah11918 commented:
Thank you! From now on, these docs should not change too much, but probably just typos etc. In that case, I will commit small fixes directly so that you do not have to be bothered. 😄

sarah11918 commented on src/content/docs/en/guides/prefetch.mdx:
```diff
@@ -114,6 +115,35 @@ To ignore slow connection detection, you can use the `ignoreSlowConnection` opti
+
+With the experimental [`clientPrerender`](https://docs.astro.build/en/reference/experimental-flags/client-prerender/) flag enabled, you can use the `eagerness` option on `prefetch()` to suggest to the browser how eagerly it should prefetch/prerender link targets.
+
+This follows the same API described in [Speculation Rules API](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/speculationrules#eagerness) and defaults to `immediate` (the most eager option). In decreasing order of eagerness, the other options are `eager`, `moderate`, and `conservative`.
```
```suggestion
This follows the same API described in the [Speculation Rules API](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/speculationrules#eagerness) and defaults to `immediate` (the most eager option). In decreasing order of eagerness, the other options are `eager`, `moderate`, and `conservative`.
```

sarah11918 commented on src/content/docs/en/guides/prefetch.mdx:
```diff
@@ -114,6 +115,35 @@ To ignore slow connection detection, you can use the `ignoreSlowConnection` opti
+
+This follows the same API described in [Speculation Rules API](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/speculationrules#eagerness) and defaults to `immediate` (the most eager option). In decreasing order of eagerness, the other options are `eager`, `moderate`, and `conservative`.
+
+The `eagerness` option allows you to balance the benefit of reduced wait times against bandwidth, memory, and CPU costs for your site visitors. Some browsers, such as Chrome, have [limits in place](https://developer.chrome.com/blog/speculation-rules-improvements#chrome-limits) to guard against over-speculating (prerendering/prefetching too many links).
```
```suggestion
The `eagerness` option allows you to balance the benefit of reduced wait times against bandwidth, memory, and CPU costs for your site visitors. Some browsers, such as Chrome, have [limits in place to guard against over-speculating](https://developer.chrome.com/blog/speculation-rules-improvements#chrome-limits)  (prerendering/prefetching too many links).
```

sarah11918 commented on src/content/docs/en/guides/prefetch.mdx:
```diff
@@ -114,6 +115,35 @@ To ignore slow connection detection, you can use the `ignoreSlowConnection` opti
+
+The `eagerness` option allows you to balance the benefit of reduced wait times against bandwidth, memory, and CPU costs for your site visitors. Some browsers, such as Chrome, have [limits in place](https://developer.chrome.com/blog/speculation-rules-improvements#chrome-limits) to guard against over-speculating (prerendering/prefetching too many links).
+
+To use `prefetch()` programmatically with large sets of links, you can set `eagerness: 'moderate'` to get advantage of [First In, First Out (FIFO)](https://en.wikipedia.org/wiki/FIFO_(computing_and_electronics)) strategies and browser heuristics to let the browser decide when to prerender/prefetch them and in what order:
```
```suggestion
To use `prefetch()` programmatically with large sets of links, you can set `eagerness: 'moderate'` to take advantage of [First In, First Out (FIFO)](https://en.wikipedia.org/wiki/FIFO_(computing_and_electronics)) strategies and browser heuristics to let the browser decide when to prerender/prefetch them and in what order:
```

sarah11918 commented on src/content/docs/en/guides/prefetch.mdx:
```diff
@@ -152,6 +182,8 @@ The most common browsers support Astro's prefetching with subtle differences:
 
 Chrome supports `<link rel="prefetch">`. Prefetching works as intended.
 
+It also fully supports `<script type="speculationrules">` from the [Speculation Rules API](https://developer.mozilla.org/en-US/docs/Web/API/Speculation_Rules_API), which can be used to futher describe [prefetching strategies and rules](#prefetch-strategies-with-eagerness-option), enhancing user experience for your Chrome users. You'll need to enable [`clientPrerender`](https://docs.astro.build/en ...
```
```suggestion
It also fully supports `<script type="speculationrules">` from the [Speculation Rules API](https://developer.mozilla.org/en-US/docs/Web/API/Speculation_Rules_API), which can be used to futher describe [prefetching strategies and rules](#eagerness), enhancing user experience for your Chrome users. You'll need to enable [`clientPrerender`](/en/reference/experimental-flags/client-prerender/) experiment to utilize this functionality with `prefetch()`
```

sarah11918 commented on src/content/docs/en/guides/prefetch.mdx:
```diff
@@ -114,6 +115,35 @@ To ignore slow connection detection, you can use the `ignoreSlowConnection` opti
+
+<p><Since v="5.6.0" /></p>
+
+With the experimental [`clientPrerender`](]/en/reference/experimental-flags/client-prerender/) flag enabled, you can use the `eagerness` option on `prefetch()` to suggest to the browser how eagerly it should prefetch/prerender link targets.
```
```suggestion
With the experimental [`clientPrerender`](/en/reference/experimental-flags/client-prerender/) flag enabled, you can use the `eagerness` option on `prefetch()` to suggest to the browser how eagerly it should prefetch/prerender link targets.
```

yanthomasdev commented on src/content/docs/en/guides/prefetch.mdx:
```diff
@@ -152,6 +182,8 @@ The most common browsers support Astro's prefetching with subtle differences:
 
 Chrome supports `<link rel="prefetch">`. Prefetching works as intended.
 
+It also fully supports `<script type="speculationrules">` from the [Speculation Rules API](https://developer.mozilla.org/en-US/docs/Web/API/Speculation_Rules_API), which can be used to futher describe [prefetching strategies and rules](#eagerness), enhancing user experience for your Chrome users. You'll need to enable [`clientPrerender`](/en/reference/experimental-flags/client-prerender/) experim ...
```
small typo
```suggestion
It also fully supports `<script type="speculationrules">` from the [Speculation Rules API](https://developer.mozilla.org/en-US/docs/Web/API/Speculation_Rules_API), which can be used to further describe [prefetching strategies and rules](#eagerness), enhancing user experience for your Chrome users. You'll need to enable [`clientPrerender`](/en/reference/experimental-flags/client-prerender/) experiment to utilize this functionality with `prefetch()`
```

yanthomasdev reviewed (commented):
Thanks, good work!

sarah11918 commented on src/content/docs/en/guides/prefetch.mdx:
```diff
@@ -114,6 +115,35 @@ To ignore slow connection detection, you can use the `ignoreSlowConnection` opti
 
+### `eagerness`
+
+<p><Since v="5.6.0" /></p>
```
```suggestion
<p>
**Type:** `'immediate' | 'eager' | 'moderate' | 'conservative'`<br />
**Default:** `'immediate'`<br />
<Since v="5.6.0" />
</p>
```

sarah11918 commented on src/content/docs/en/guides/prefetch.mdx:
```diff
@@ -114,6 +115,57 @@ To ignore slow connection detection, you can use the `ignoreSlowConnection` opti
+prefetch('/getting-started'); // defaults to `{ eagerness: 'immediate' }`
+
+// This page may not be visited
+prefetch('/our-history', { eagerness: 'moderate' });
```
```suggestion
prefetch('/terms-of-service', { eagerness: 'moderate' });
```

sarah11918 commented:
Would love a final approval from @Marocco2 here, as I added some stuff, so make sure this all still seems correct!

@ArmandPhilippot , I added the API reference at the top so couldn't hurt for you to confirm. 😄 

Then, I think this is ready to merge into the 5.6.0 PR branch for this week's release! 🥳

Marocco2 commented:
> Would love a final approval from @Marocco2 here, as I added some stuff, so make sure this all still seems correct!
> 
> @ArmandPhilippot , I added the API reference at the top so couldn't hurt for you to confirm. 😄 
> 
> Then, I think this is ready to merge into the 5.6.0 PR branch for this week's release! 🥳 

I've checked. All good 👍

sarah11918 commented:
@ArmandPhilippot It's because previews aren't generated for PRs not to the main branch! (I forget sometimes myself.) We can double double check once it's merged into the 5.6.0 PR, because *that* is made to the main branch. I switched target branches earlier than usual, because we're ahead of the game, but I forgot (myself) that we lose previews that way.
